### DISCUSSION TITLE
Feature request: Require approval from specific assigned reviewers (Required + Optional reviewers)

### DISCUSSION THREAD

**Discussion by WoodyWoodsta** [community/community#38668] [Pull Requests]
There seems to be no way to get something like this to work with the protection rules and CODEOWNERS file as it stands, at least that I know about.

## The Problem

At face value, the problem is enabling auto-merge on a PR which has multiple reviewers (and a branch protection rule to require at least `x` approvals) will merge automatically after the first `x` approval(s) which may not be as intended.

Generally speaking, the problem is that there is no way for the PR author to specify specific required reviewers for their PR. In these cases, a simple "number of required reviewers" branch protection rule doesn't help.

Has this been approached before?

## Proposed Solution

Azure Repos has the concept of required and optional reviewers, and a rule to require all required reviewers to approve before merge. This seems like a reasonable solution which covers most, if not all, use cases.

**Comment by stefangimmillaro** [comment]
glad to find someone else already requested this feature

I would really like this too. I would like to specify automerge conditions on a specific pull request:
- Approval from specific reviews
- automerge only within a certain time of day window

lower priority for me personally
- number of approvals
- automerge only after another specific pr merges

**Comment by mbs-c** [comment]
We would love this feature, for slightly different reasons. Our ideal development workflow is as follows:

* Develop code change
* Create PR
* Assign reviewers (optional, depending on the type of change)
* Write your squash commit message and enable automerge
* Move on to the next task

GitHub currently doesn't allow us to follow this workflow, since you cannot assign specific people as required reviewers to specific PRs. Instead, you have to watch your emails and wait for the approval before you can write the squash commit message and click the merge button, which causes unnecessary context switches. I want to be able to request mandatory reviews from specific reviewers in a PR, without branch protection rules requiring x reviews for every PR.

**Comment by lukepighetti** [comment]
ran into this use case when trialing graphite for stacked diffs. we don't want to impose a requirement for approvals, but we do want to assign reviewers as needed and auto-merge when approval is received

**Comment by MuhammadAlZafarKhan** [comment]
This is a common missing feature in standard GitHub Pull Request (PR) settings.

**The workarounds include:**

- The CODEOWNERS File: You define users or teams in a .github/CODEOWNERS file who are responsible for specific files or directories. When you enable the "Require review from Code Owners" branch protection rule, a PR that modifies a file with a designated owner must be approved by at least one of those owners before it can be merged.

- Implementing Custom GitHub Actions: The Action runs on the pull_request or pull_request_review event. It checks the list of assigned reviewers and compares it to the list of approving reviewers. If an assigned reviewer hasn't approved, the Action sets a failing status check on the PR. You then enable a Branch Protection Rule to require that specific status check to pass before merging. The PR will remain blocked until the Action reports success, which only happens once all specific reviewers have approved.

- Using a Third-Party Tool: These tools provide a layer of protection that allows for fine-grained control over merge conditions, such as:
1. Requiring approvals from specific teams or roles.
2. Conditional approvals based on file paths or change types.

**Comment by alamenai** [comment]
I migrated from Azure Devops which has this feature then ended with GitHub does not have it!

**Reply by JVApen** [reply]
The same, it's a huge gap coming from Azure DevOps, so sad that after 4 years, this is still not implemented
